### ISSUE TITLE
no-unused-variable does not recognize destructuring assignment

### ISSUE THREAD

**Issue by jnu** [palantir/tslint#1127]
### Bug Report
- **TSLint version**: `3.6.0`
- **TypeScript version**: `1.8.9`
- **Running TSLint via**: SublimeLinter
#### TypeScript code being linted

``` ts
class Foo {

  private bar: string = 'bar';

  public method(): void {
    const { bar } = this;
    console.log(bar);
  }

}
```

with `tslint.json`:

``` json
"no-unused-variable": true
```
#### Actual behavior

`Error: unused variable: 'bar'`
#### Expected behavior

should detect that `bar` is used through destructuring assignment

**Comment by JKillian** [comment]
Hi @jnu, the warning comes because of the line:

```
private bar: string = 'bar';
```

That `bar`, is in fact, unused. ;)

**Comment by jnu** [comment]
what? yes, it is used.

compare the semantically equivalent:
```
class Foo {

  private bar: string = 'bar';

  public method(): void {
    console.log(this.bar);
  }

}
```

there is, correctly, no lint error in this case. the linter appears not to be catching that `bar` is used through [destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment).

**Comment by JKillian** [comment]
Whoops, my bad! I just skimmed your code for `this.bar` and wasn't really paying attention to how you were in fact using it...

**Comment by jnu** [comment]
:) thanks!

**Comment by JKillian** [comment]
So this is actually a somewhat complicated issue... the tricky thing there is that the `bar` in `const { bar } = this` actually refers to two things: the identifier of the field on `this` and the identifier of the local variable it creates.

To determine if a variable is used, we use an API of TypeScript that finds all the references to an identifier. (This is the same API that atom-ts or vscode uses when you select a variable and choose something like "find references").

This is a confusing scenario, because it can be unclear what should happen. And in fact, there are many TS issues about similar situations:

https://github.com/Microsoft/TypeScript/issues/6459
https://github.com/Microsoft/TypeScript/issues/6312
etc.

There's been a recent PR that loos helpful and suggests that this situation may have been fixed: https://github.com/Microsoft/TypeScript/pull/7945

Essentially this is an issue that we just have to hope changes in TS, which will fix things here in TSLint. If there's a definite bug in TS (even given the changes above), I'll file an issue with them.

**Comment by jnu** [comment]
great, thanks for looking into it!

**Comment by andy-hanson** [comment]
Should be fixed by #2235.
